### PR TITLE
Extend support for contraction 6D

### PR DESCRIPTION
**Pull request by bartekxk**

### CODE REVIEW

File: profiler/src/profile_contraction_bilinear.cpp [hunk 1]
```diff
@@ -100,7 +100,9 @@ int profile_contraction_bilinear(int argc, char* argv[])
                 assign_default_strides(cde_layout, StridesE, {M[0], M[1], N[0], N[1]});
                 assign_default_strides(cde_layout, StridesD, {M[0], M[1], N[0], N[1]});
             }
-            bool pass = ck::profiler::profile_contraction_impl<ALayout,
+            constexpr ck::index_t NumDim = 2;
```
Comment (aosewski) [inline profiler/src/profile_contraction_bilinear.cpp]: Isn't it possible to profile also 6D cases ?

File: profiler/src/profile_contraction_scale.cpp [hunk 2]
```diff
@@ -99,7 +99,9 @@ int profile_contraction_scale(int argc, char* argv[])
                 assign_default_strides(cde_layout, StridesD, {M[0], M[1], N[0], N[1]});
             }
 
-            bool pass = ck::profiler::profile_contraction_impl<ALayout,
+            constexpr ck::index_t NumDim = 2;
```
Comment (aosewski) [inline profiler/src/profile_contraction_scale.cpp]: Isn't it possible to profile also 6D cases ?

File: test/contraction/test_contraction.cpp [hunk 3]
```diff
@@ -122,17 +130,31 @@ TYPED_TEST_SUITE(TestContractionScale, ScaleKernelTypes);
 TYPED_TEST(TestContractionBilinear, bilinear)
 {
     this->p_cd_element_op = std::make_unique<Bilinear>(1.f, 1.f);
```
Comment (aosewski) [inline test/contraction/test_contraction.cpp]: Please use coefficients other than `1.f`

File: test/contraction/test_contraction.cpp [hunk 4]
```diff
@@ -122,17 +130,31 @@ TYPED_TEST_SUITE(TestContractionScale, ScaleKernelTypes);
...
-    this->Run();
+    this->template Run<6>({{2, 3, 2, 3, 2, 3}, {2, 3, 2, 3, 2, 3}, {2, 2, 2, 2, 2, 4}});
+    this->template Run<6>({{1, 1, 1, 3, 2, 3}, {1, 1, 1, 3, 2, 3}, {1, 1, 1, 2, 2, 4}});
+    this->template Run<2>({{16, 16}, {16, 16}, {16, 16}});
+    this->template Run<2>({{8, 8}, {16, 16}, {8, 8}});
```
Comment (aosewski) [inline test/contraction/test_contraction.cpp]: Please use different values for all Ms, Ns, and Ks here and in all other test cases.

File: test/contraction/test_contraction.cpp [hunk 5]
```diff
@@ -122,17 +130,31 @@ TYPED_TEST_SUITE(TestContractionScale, ScaleKernelTypes);
...
 }
 
 TYPED_TEST(TestContractionScale, scale)
 {
     this->p_cd_element_op = std::make_unique<Scale>(1.f);
```
Comment (aosewski) [inline test/contraction/test_contraction.cpp]: Please use coefficients other than `1.f`

File: test/contraction/test_contraction.cpp [hunk 6]
```diff
@@ -165,15 +187,29 @@ TYPED_TEST_SUITE(TestContractionScaleMixedPrecision, ScaleKernelTypesMixedPrecis
 TYPED_TEST(TestContractionBilinearMixedPrecision, bilinear)
 {
     this->p_cd_element_op = std::make_unique<Bilinear>(1.f, 1.f);
```
Comment (aosewski) [inline test/contraction/test_contraction.cpp]: Please use coefficients other than 1.f

File: test/contraction/test_contraction.cpp [hunk 7]
```diff
@@ -165,15 +187,29 @@ TYPED_TEST_SUITE(TestContractionScaleMixedPrecision, ScaleKernelTypesMixedPrecis
...
-    this->Run();
+    this->template Run<6>({{2, 3, 2, 3, 2, 3}, {2, 3, 2, 3, 2, 3}, {2, 2, 2, 2, 2, 4}});
+    this->template Run<6>({{1, 1, 1, 3, 2, 3}, {1, 1, 1, 3, 2, 3}, {1, 1, 1, 2, 2, 4}});
+    this->template Run<2>({{16, 16}, {16, 16}, {16, 16}});
+    this->template Run<2>({{8, 8}, {16, 16}, {8, 8}});
 }
 
 TYPED_TEST(TestContractionScaleMixedPrecision, scale)
 {
     this->p_cd_element_op = std::make_unique<Scale>(1.f);
```
Comment (aosewski) [inline test/contraction/test_contraction.cpp]: Please use coefficients other than 1.f

**Review by aosewski — COMMENTED** [review]
Just a few minor comments. Other than that looks good.

File: test/contraction/test_contraction.cpp (hunk 3 above)
Comment (bartekxk) [reply test/contraction/test_contraction.cpp]: There are also tests with other coefficients

File: test/contraction/test_contraction.cpp (hunk 5 above)
Comment (bartekxk) [reply test/contraction/test_contraction.cpp]: There are also tests with other coefficients

File: test/contraction/test_contraction.cpp (hunk 6 above)
Comment (bartekxk) [reply test/contraction/test_contraction.cpp]: There are also tests with other coefficients

Comment (bartekxk) [reply test/contraction/test_contraction.cpp]: There are also tests with other coefficients

File: test/contraction/test_contraction.cpp (hunk 7 above)
Comment (bartekxk) [reply test/contraction/test_contraction.cpp]: There are also tests with other coefficients

File: profiler/src/profile_contraction_bilinear.cpp (hunk 1 above)
Comment (bartekxk) [reply profiler/src/profile_contraction_bilinear.cpp]: Done.

File: profiler/src/profile_contraction_scale.cpp (hunk 2 above)
Comment (bartekxk) [reply profiler/src/profile_contraction_scale.cpp]: Done.

File: test/contraction/test_contraction.cpp (hunk 4 above)
Comment (bartekxk) [reply test/contraction/test_contraction.cpp]: Done.

File: profiler/README.md [hunk 8]
```diff
@@ -49,24 +49,30 @@ Best Perf: 1.42509 ms, 102.988 TFlops, 234.086 GB/s
 
 ## Profile contraction kernels
 ```bash
-#arg1: tensor operation (contraction_bilinear=CONTRACTION+Bilinear)
+#arg1: tensor operation ( OP_NAME :  OP_DESC )
```
Comment (aosewski) [inline profiler/README.md]: Why you have put here generic names ?

File: profiler/README.md [hunk 9]
```diff
@@ -49,24 +49,30 @@ Best Perf: 1.42509 ms, 102.988 TFlops, 234.086 GB/s
...
-#arg17 to 32: Strides for A, B, D and E (skip for default)
+#arg4: Number of dimension for M, N and K (one for all)
+#arg5: matrix layout (0: A[m0, m1, k0, k1] * B[k0, k1, n0, n1] + 
+#                 D[m0, m1, n0, n1] = E[m0, m1, n0, n1];
+#                     1: A[m0, m1, k0, k1] * B[n0, n1, k0, k1] + 
+#                 D[m0, m1, n0, n1] = E[m0, m1, n0, n1];
+#                     2: A[k0, k1, m0, m1] * B[k0, k1, n0, n1] + 
+#                 D[m0, m1, n0, n1] = E[m0, m1, n0, n1];
+#                     3: A[k0, k1, m0, m1] * B[n0, n1, k0, k1] + 
+#                 D[m0, m1, n0, n1] = E[m0, m1, n0, n1])
```
Comment (aosewski) [inline profiler/README.md]: Earlier It was more nicely formatted.

**Review by LisaDelaney — APPROVED** [review]

**Review by zjing14 — DISMISSED** [review]

**Review by zjing14 — APPROVED** [review]

**Review by lpaoletti — APPROVED** [review]
Looked over the content in the Docs folder. 
However, Lisa already reviewed.

**Pull request merged**